teachertool/src/components/makecodeFrame.tsx

```diff
@@ -0,0 +1,138 @@
+}
+
+type FrameState = "loading" | "no-project" | "opening-project" | "project-open" | "closing-project";
+
+interface PendingMessage {
+    original: pxt.editor.EditorMessageRequest;
+    handler: (response: any) => void;
+}
+
+export const MakeCodeFrame: React.FC<MakeCodeFrameProps> =
+    ( { pageSourceUrl,
+        highContrast,
+        onFrameClose,
+        onFrameOpen
+    } ) => {
+    let ref: HTMLIFrameElement | undefined;
+    const messageQueue: pxt.editor.EditorMessageRequest[] = [];
+    let nextId: number = 0;
+    let pendingMessages: {[index: string]: PendingMessage} = {};
+    const [frameState, setFrameState] = useState<FrameState>("loading");
+    const [loadPercent, setLoadPercent] = useState(0);
+    const [workspaceReady, setWorkspaceReady] = useState(false);
+
+    let frameRef = useRef<HTMLIFrameElement>(null);
+
+    useEffect(() => {
+        // logic we want to do when the iframe is loaded
+        if (ref && ref.contentWindow) {
+            window.addEventListener("message", onMessageReceived);
+            ref.addEventListener("load", handleFrameReload)
+
+        }
+
+        // logic we want when the iframe unmounts
+        return () => {
+            window.removeEventListener("message", onMessageReceived);
+        }
+    }, [])
+
+    useEffect(() => {
+        sendMessageAsync({
+            type: "pxteditor",
+            action: "sethighcontrast",
+            on: highContrast
+        }  as pxt.editor.EditorMessageSetHighContrastRequest);
+    }, [highContrast])
+
+
+    useEffect(() => {
+        if (frameState === "project-open" /*&& this.props.save*/) {
+            setFrameState("closing-project");
+            onFrameClose();
+        }
+        else if (frameState === "no-project") {
+            setFrameState("opening-project");
+            onFrameOpen();
+        }
+    }, [frameState]);
+
+
+    const handleFrameReload = () => {
+        setFrameState("loading")
+    }
+
+    const onMessageReceived = (event: MessageEvent) => {
+        const data = event.data as pxt.editor.EditorMessageRequest;
+        if (frameState === "opening-project") setLoadPercent(Math.min((loadPercent || 0) + 7, 95));
+
+        if (data.type === "pxteditor" && data.id && pendingMessages[data.id]) {
+            const pending = pendingMessages[data.id];
+            pending.handler(data);
+            delete pendingMessages[data.id];
+            return;
+        }
+
+        switch (data.action) {
+            case "newproject":
+                if (!workspaceReady) {
+                    setWorkspaceReady(true);
+                    sendMessageAsync(); // Flush message queue
+                }
+                if (frameState === "loading") {
+                    setFrameState("no-project");
+                }
+                break;
+            default:
+                console.log(JSON.stringify(data, null, 4));
+        }
+    }
+
+    const sendMessageAsync = (message?: any) => {
```

Comment:
Something to think about is how will we actually call into this function when we want to send a message? I could certainly be missing something, but at the moment I don't think it's exposed in any way, and I'm not sure it'll be trivial to do so without coupling this class quite tightly to Teacher Tool state.

Did you already have a specific plan in mind around how we'll call this, or is it still TBD?

Reply:
Yeah, that's a good point. It's something that I haven't thought about much, so definitely still TBD. I'll play with this a bit to see what we can do.